### PR TITLE
Add megatron/core/ops: an implementation home for kernel backends

### PR DESCRIPTION
**Pull request by guihong-nv**
## What

Adds `megatron/core/ops` and `megatron/core/inference/ops`: one package per operation family, holding the backends for that operation side by side, plus the contract they meet.

**This is an implementation home, not a selection layer.** There is no registry and no resolver; nothing here decides which backend a run gets. That stays with the `BackendSpecProvider` implementations, which import their targets from here.

**Nothing calls it yet** — 21 files, +1266, **−0**. It cannot change behaviour and can be reviewed on its own. #6763 switches the callers over.

## Shape

| file | holds |
|---|---|
| `<family>/__init__.py` | the contract every backend for this family must meet |
| `<family>/backends.py` | every implementation, side by side |

Backends are named for the family and the implementation — `NormTE`, `NormApex`, `NormTorch`, `LinearTE`, `MoeLocal` — so the class name alone says what it is. They sit together on purpose: a family's backends answer the same question and are read by comparison.

## The one rule

Every optional-package import lives **inside the method that returns its target**, never at module scope:

```python
class NormTE:
    REQUIRES = "transformer_engine"
    DETERMINISM = "unknown"

    def layer_norm(self, rms_norm=False, for_qk=False, has_residual=False):
        from megatron.core.extensions.transformer_engine import TENorm   # here, not above
        return TENorm
```

So importing `megatron.core.ops` pulls in no optional dependency and no backend nobody selected — pinned by `tests/unit_tests/ops/test_import_hygiene.py`.

That is what lets a provider name a backend without the **call site** having to guard on whether it is installed, which is what the scattered `HAVE_TE` / `HAVE_APEX` flags were doing. Removing those is #6763.

## Tracing which implementation a run gets

Open the provider. `TESpecProvider.layer_norm` names `NormTE`; `NormTE.layer_norm` names `TENorm`. Two hops, no globals, no dependence on import order or on what happens to be installed.

For comparison, the same question on `main` takes five hops and still depends on whether Apex imported and whether an earlier call mutated `LNImpl` — a module-level global that eight files each define their own copy of.

## Testing

`tests/unit_tests/ops/` — 8×GPU, 20 passed. Backend targets (each backend returns the exact class it did before the move), cross-entropy parity against `F.cross_entropy`, and import hygiene.

See `megatron/core/ops/README.md`.

### CODE REVIEW

**Comment by guihong-nv** [comment]
Closing: nothing imports this.

#6763 was reworked so that the providers keep importing their targets from where those targets already live, rather than from `megatron/core/ops`. That makes #6763 self-contained and achieves milestone 1 — enforce `BackendSpecProvider`, remove the `HAVE_*` flags — without this package existing at all.

Which leaves this PR adding ~1300 lines that nothing calls. "Don't merge code nothing calls" seems like the right bar, so the folder should come back with its first real user: the milestone that actually moves implementations into it (`Move norm, activation, and loss kernels`, and the ones after it). The contracts and the backend classes here are a reasonable starting point when that happens; the branch stays on the fork.

One thing worth carrying forward from it — the rule that made the `HAVE_*` cleanup possible in the first place: **an optional-package import belongs inside the method that returns its target, never at module scope.** That is what lets a provider name a backend without the call site having to guard on whether it is installed. #6763 follows it.
